### PR TITLE
Add Mypy to the CI

### PR DESCRIPTION
**Pull request by Fokko**
Mypy can be used to validate the types of the functions and inheritance.

### CODE REVIEW

**Comment by Fokko** [comment]
cc @jun-he

File: python/iceberg/api/filterable.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 
 class Filterable(object):
-    ALL_COLUMNS = ("*",)
+    ALL_COLUMNS = ["*"]
```
Comment (rdblue) [inline python/iceberg/api/filterable.py:20]: This seems like an odd change to me. Why wouldn't we use an immutable value?

Comment (Fokko) [reply python/iceberg/api/filterable.py:20]: We use tuples and list mixed together throughout the codebase:
https://github.com/apache/incubator-iceberg/blob/57b10995aade1362e582cef68b20014023556501/python/iceberg/core/manifest_reader.py#L36

I find a list more appropriate than a tuple for the columns.

Comment (rdblue) [reply python/iceberg/api/filterable.py:20]: For constants, I'd prefer to use a value that can't be modified accidentally.

File: python/iceberg/core/base_metastore_tables.py [hunk 2]
```diff
@@ -50,4 +50,4 @@ def default_warehouse_location(self, conf, database, table):
         if warehouse_location is None:
             raise RuntimeError("Warehouse location is not set: hive.metastore.warehouse.dir=null")
 
-        return "%s/%s.db/%s".format(warehouse_location, database, table)
+        return f"{warehouse_location}/{database}.db/{table}"
```
Comment (rdblue) [inline python/iceberg/core/base_metastore_tables.py:53]: Was this required by Mypy? Looks more like a style change.

File: python/iceberg/core/data_table_scan.py [hunk 3]
```diff
@@ -37,9 +37,9 @@
 
 class DataTableScan(BaseTableScan):
 
-    SNAPSHOT_COLUMNS = ("snapshot_id", "file_path", "file_ordinal", "file_format", "block_size_in_bytes",
+    SNAPSHOT_COLUMNS = ["snapshot_id", "file_path", "file_ordinal", "file_format", "block_size_in_bytes",
```
Comment (rdblue) [inline python/iceberg/core/data_table_scan.py:40]: I think this should continue to be an immutable tuple.

File: python/iceberg/api/partition_spec.py [hunk 4]
```diff
@@ -231,7 +231,7 @@ def identity(self, source_name):
         return self
 
     def year(self, source_name):
-        name = "{}_year".format(source_name)
+        name = f"{source_name}_year"
```
Comment (rdblue) [inline python/iceberg/api/partition_spec.py:234]: Was this required by the analysis tool?

File: python/iceberg/core/base_metastore_tables.py (hunk 2 above)
Comment (Fokko) [reply python/iceberg/core/base_metastore_tables.py:53]: The thing is actually not templated:
```
MacBook-Pro-van-Fokko:incubator-iceberg fokkodriesprong$ python3
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 26 2018, 19:50:54) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> '%s'.format('vo')
'%s'
>>> '{}'.format('vo')
'vo'
```

**Comment by Fokko** [comment]
@rdblue I tried to fix this as well, without using `pre-commit` and fix it using tox. I've noticed that there is an overlap with https://github.com/apache/incubator-iceberg/pull/1034. Could you merge that one first? I'll fix this one afterward.

**Comment by rdblue** [comment]
@Fokko, I merged #1034, so you should be unblocked here. Thanks for working on this and for moving it to Tox!

**Pull request closed without merging**